jupyterhub_traefik_proxy/toml.py

```diff
@@ -0,0 +1,275 @@
+"""Traefik implementation
+
+Custom proxy implementations can subclass :class:`Proxy`
+and register in JupyterHub config:
+
+.. sourcecode:: python
+
+    from mymodule import MyProxy
+    c.JupyterHub.proxy_class = MyProxy
+
+Route Specification:
+
+- A routespec is a URL prefix ([host]/path/), e.g.
+  'host.tld/path/' for host-based routing or '/path/' for default routing.
+- Route paths should be normalized to always start and end with '/'
+"""
+
+# Copyright (c) Jupyter Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+import json
+import os
+import asyncio
+import toml
+
+from traitlets import Any, default, Unicode
+
+from . import traefik_utils
+from jupyterhub.proxy import Proxy
+from jupyterhub_traefik_proxy import TraefikProxy
+
+
+class TraefikTomlProxy(TraefikProxy):
+    """JupyterHub Proxy implementation using traefik and toml config file"""
+
+    mutex = Any()
+
+    @default("mutex")
+    def _default_mutex(self):
+        return asyncio.Lock()
+
+    toml_static_config_file = Unicode(
+        "traefik.toml", config=True, help="""traefik's static configuration file"""
+    )
+
+    toml_dynamic_config_file = Unicode(
+        "rules.toml", config=True, help="""traefik's dynamic configuration file"""
+    )
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        self.routes_cache = {}
+
+    async def _setup_traefik_static_config(self):
+        await super()._setup_traefik_static_config()
+
+        self.static_config["file"] = {"filename": "rules.toml", "watch": True}
+        toml_string = toml.dumps(self.static_config)
+
+        try:
+            with open(self.toml_static_config_file, "w") as f:
+                f.write(toml_string)
+            open(self.toml_dynamic_config_file, "a").close()
+        except IOError:
+            self.log.exception("Couldn't set up traefik's static config.")
+        except:
+            self.log.error(
+                "Couldn't set up traefik's static config. Unexpected error:",
+                sys.exc_info()[0],
+            )
+
+    def _start_traefik(self):
+        self.log.info("Starting traefik...")
+        try:
+            self._launch_traefik(config_type="toml")
+        except FileNotFoundError as e:
+            self.log.error(
+                "Failed to find traefik \n"
+                "The proxy can be downloaded from https://github.com/containous/traefik/releases/download."
+            )
+            raise
+
+    def _clean_resources(self):
+        try:
+            os.remove(self.toml_static_config_file)
```

Comment:
We'll need to be careful here: we should only delete the 'static' config file if we are writing it (i.e. `if self.should_start`)